### PR TITLE
fix #7175: Trigger beforePopState event before popState event when page is ssr

### LINKED ISSUE #7175: beforePopState() does not fire at all
# Bug report

## Describe the bug

When the user clicks on their browser's back button, I want to execute some custom logic for my app. I tried using Router.beforePopState() in my React component, but it never gets called. My goal is to be able to detect when the back button is pressed. I found example code in #6784  that shows the behaviour that I am talking about.

## To Reproduce
Repo: https://github.com/mlnor27/nextjs-bug-to-reproduce
Steps to reproduce the behavior, please provide code snippets or a repository:
1. go to http://localhost:3000/
2. manually type out the link to the about page in your browser's url bar and press enter: http://localhost:3000/about
3. click your browser's back button
4. there is no alert popup, beforePopState never seems to get called

You can see that pop up if you click the link to 'about' page from the '/', instead of manually typing it into the url bar as in step 2.

## Expected behavior
I want to detect when the user clicks their back button. In the provided example I expect to see an alert popping up.

## Screenshots
Here is a screen capture:
1. manually typing into the address bar results in no popups
2. clicking the link, then pressing the back button to show that popup
about

## System information
 - OS: Linux
 - Browser (if applies) Firefox Version 64.0.2 (64-bit)
 - Version of Next.js: 8.1.0

### PR DESCRIPTION
**Pull request by nblthree**
I decided to make this PR to check the Tests https://github.com/zeit/next.js/issues/7175#issuecomment-487199811

### CODE REVIEW

**Comment by lfades** [comment]
@MarchWorks Did you check if there are no unexpected secondary effects ? like the one being described in this comment:
```
// Make sure we don't re-render on initial load,
// can be caused by navigating back from an external site
```

**Comment by timneutkens** [comment]
@ijjk should review this

File: packages/next-server/lib/router/router.ts [hunk 1]
```diff
@@ -105,18 +105,18 @@ export default class Router implements BaseRouter {
...
-
     // If the downstream application returns falsy, return.
     // They will then be responsible for handling the event.
     if (this._bps && !this._bps(e.state)) {
       return
     }
 
+    // Make sure we don't re-render on initial load,
+    // can be caused by navigating back from an external site
+    if (e.state.options && e.state.options.fromExternal) {
```
Comment (MarchWorks) [inline packages/next-server/lib/router/router.ts]: Why `e.state.options.fromExternal` always return undefined I made sure that the `unload` event is being triggered when visiting an external site but it end up being undefined when navigating back
PS: Doesn't have anything to do with the changes that I made

**Comment by MarchWorks** [comment]
@lfades I can't think of any other check than the one that you have mentioned and there is no problem with it logically ( check the above review to know why I wrote 'logically' )

**Comment by MarchWorks** [comment]
@lfades logically it won't cause any problem

**Comment by ijjk** [comment]
@MarchWorks this doesn't appear to solve the mentioned issue. It looks like the issue is describing wanting to be able to capture popState when a user manually navigates. This doesn't work since popState isn't fired in the case they describe. If you follow the reproduction steps from the issue using this PR the alert still doesn't show. 

I replied to the original thread [here](https://github.com/zeit/next.js/issues/7175#issuecomment-487330511) explaining this.

**Comment by MarchWorks** [comment]
@ijjk I am sure that it was working in chrome (wasn't working in firefox) I think that when I was testing the last time there wasn't a ssr after navigating back from `about.js` then I will be closing this. just a little question doesn't `e.state.options.fromExternal` always return undefined

**Comment by ijjk** [comment]
> doesn't e.state.options.fromExternal always return undefined

This is set [here](https://github.com/zeit/next.js/blob/2eee87663057b98f73c21944b15012803286b4a5/packages/next-server/lib/router/router.ts#L82) when navigating away. It is true when coming back from an external site in a browser that fires a stale popState on initial load like safari and older versions of chrome

**Pull request closed without merging**